lua/luasnip/util/environ.lua

```diff
@@ -1,190 +1,193 @@
-local util = require("luasnip.util.util")
-
-local eager_vars = {
-	["TM_CURRENT_LINE"] = true,
-	["TM_CURRENT_WORD"] = true,
-	["TM_LINE_INDEX"] = true,
-	["TM_LINE_NUMBER"] = true,
-	["TM_SELECTED_TEXT"] = true,
-	["SELECT_RAW"] = true,
-	["SELECT_DEDENT"] = true,
-}
--- These are the vars that have to be populated once the snippet starts to avoid any issue
-local function _fill_eager_vars(env, pos)
-	env.TM_CURRENT_LINE =
-		vim.api.nvim_buf_get_lines(0, pos[1], pos[1] + 1, false)[1]
-	env.TM_CURRENT_WORD = util.word_under_cursor(pos, env.TM_CURRENT_LINE)
-	env.TM_LINE_INDEX = tostring(pos[1])
-	env.TM_LINE_NUMBER = tostring(pos[1] + 1)
-	env.SELECT_RAW, env.SELECT_DEDENT, env.TM_SELECTED_TEXT =
-		util.get_selection()
-end
+local builtin_vars = require("luasnip.util._builtin_vars")
 
-local lazy_vars = {}
+local function tbl_to_lazy_env(tbl)
+	local function wrapper(varname)
+		local val_ = tbl[varname]
+		if type(val_) == "function" then
+			return val_()
+		end
+		return val_
+	end
 
-local Environ = {}
-function Environ:new(pos, o)
-	o = o or {}
-	setmetatable(o, self)
-	_fill_eager_vars(o, pos)
-	return o
+	return wrapper
 end
 
-function Environ:__index(key)
-	local v = lazy_vars[key]()
-	rawset(self, key, v)
-	return v
-end
+local namespaces = {
+	[""] = {
+		init = builtin_vars.eager,
+		vars = tbl_to_lazy_env(builtin_vars.lazy),
+		eager = {},
+		is_table = builtin_vars.is_table,
+	},
+}
 
-function Environ:override(env, new_env)
-	for k, v in pairs(new_env) do
-		env[k] = v
+-- Namespaces allow users to define their own environmet variables
+local function _resolve_namespace_var(full_varname)
```

Comment:
How are lazy variables from the "" namespace resolved right now? In case they're not, checking them before trying other namespaces seems the right approach.

Reply:
They are checked after because they can't be overridden so it's safe.

Reply:
Ahh, L37, I'm blind :sweat_smile: